# pandas

Now that we know how to get our data in, let's look a little more at `pandas` and what it can do. To do this, let's bring in some data from Zillow.

The data frame we will be working with contains data from Zillow.com, the housing web site, on the number of homes, the median value of homes, etc. at-risk for flooding due to rising sea levels.

You can find many different Zillow data sets [here](https://www.zillow.com/research/data/). The “uw” here stands for “under water” in a very literal sense. I first thought they were talking about mortgages!

In [1]:
import numpy as np
import pandas as pd

uw = pd.read_csv('https://github.com/aaiken1/fin-data-analysis-python/raw/main/data/zestimatesAndCutoffs_byGeo_uw_2017-10-10_forDataPage.csv')

## Looking at our data

Let's peak at this data. You can do that below, or using the `.info` method.

In [2]:
uw.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2610 entries, 0 to 2609
Data columns (total 24 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   RegionType                    2610 non-null   object 
 1   RegionName                    2610 non-null   object 
 2   StateName                     2609 non-null   object 
 3   MSAName                       1071 non-null   object 
 4   AllHomes_Tier1                2610 non-null   float64
 5   AllHomes_Tier2                2610 non-null   float64
 6   AllHomes_Tier3                2610 non-null   float64
 7   AllHomes_AllTiers             2610 non-null   float64
 8   UWHomes_Tier1                 2610 non-null   float64
 9   UWHomes_Tier2                 2610 non-null   int64  
 10  UWHomes_Tier3                 2610 non-null   float64
 11  UWHomes_AllTiers              2610 non-null   float64
 12  UWHomes_TotalValue_Tier1      2610 non-null   float64
 13  UWH

There's a concept in data analysis called **tidy data**. The characteristics of tidy data are:

- Each variable forms a column.
- Each observation forms a row.
- Each type of observational unit forms a table.

```{figure} ../images/04-tidy.png
---
name: 04-tidy.png
align: center
---
Tidy data.
```

The first thing to note about this data is that there are several tiers. Some numbers are at the national level, some at the region, some at the state, and some at the MSA. **Tidy data** refers to one row, one observation, across different variables (columns) and, possibly, across time. When you have multiple observations of the same firm, state, person, etc. over time, we call that panel data. Panel data is tidy too.

This data isn’t tidy! Note that some RegionTypes (e.g. Nation, State) are actually summaries of other rows. This is bit confusing and we need to be mindful of this different levels when using the data.

Before we get to that, just a few more methods to look at the data.

First, let's look at the first five rows.

In [3]:
uw.head()

RegionType     RegionName    StateName MSAName  AllHomes_Tier1  \
0     Nation  United States          NaN     NaN      35461549.0   
1      State        Alabama      Alabama     NaN        546670.0   
2      State     California   California     NaN       3060171.0   
3      State    Connecticut  Connecticut     NaN        333904.0   
4      State       Delaware     Delaware     NaN        102983.0   

   AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
0      35452941.0      35484532.0        106399022.0       594672.0   
1        520247.0        491300.0          1558217.0         2890.0   
2       3076238.0       3169584.0          9305993.0         8090.0   
3        334688.0        336254.0          1004846.0         4434.0   
4        127573.0        127983.0           358539.0         4105.0   

   UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  UWHomes_TotalValue_AllTiers  \
0         542681  ...              5.974490e+11                 9.159540e+11   
1           2766  ...              2.583039e+09                 3.578203e+09   
2          14266  ...              2.666638e+10                 4.449536e+10   
3           3807  ...              8.689480e+09                 1.152170e+10   
4           2574  ...              1.498013e+09                 2.801847e+09   

   UWHomes_MedianValue_AllTiers  AllHomes_Tier1_ShareUW  \
0                      310306.0                0.016769   
1                      270254.0                0.005287   
2                      860373.0                0.002644   
3                      442036.0                0.013279   
4                      241344.5                0.039861   

   AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  AllHomes_AllTiers_ShareUW  \
0                0.015307                0.020458                   0.017512   
1                0.005317                0.010271                   0.006868   
2                0.004637                0.005611                   0.004313   
3                0.011375                0.021677                   0.015455   
4                0.020177                0.025675                   0.027793   

   UWHomes_ShareInTier1  UWHomes_ShareInTier2  UWHomes_ShareInTier3  
0              0.319149              0.291246              0.389605  
1              0.270043              0.258456              0.471501  
2              0.201550              0.355415              0.443035  
3              0.285512              0.245138              0.469350  
4              0.411942              0.258304              0.329754  

[5 rows x 24 columns]

`NaN` is how `pandas` shows missing values. So, that first row is `RegionType == 'Nation'` and `RegionName == 'UnitedStates'`. `StateName` and `MSAName` are undefined, since this row is **all of the data added up**. We usually don't want to have summary statistics in the same DataFrame as underlying data! What a mess!

We can also look at the last five rows.

In [4]:
uw.tail()

RegionType RegionName   StateName MSAName  AllHomes_Tier1  \
2605        Zip      98592  Washington     NaN           464.0   
2606        Zip      98595  Washington     NaN           558.0   
2607        Zip      98612  Washington     NaN           365.0   
2608        Zip      32081     Florida     NaN          1318.0   
2609        Zip      33578     Florida     NaN          5975.0   

      AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  UWHomes_Tier1  \
2605           470.0           493.0             1427.0            8.0   
2606           571.0           598.0             1727.0          119.0   
2607           376.0           409.0             1150.0           91.0   
2608          1328.0          1350.0             3996.0           91.0   
2609          4537.0          3232.0            13744.0            8.0   

      UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  \
2605             29  ...                73407109.0   
2606            104  ...                13761194.0   
2607             79  ...                14406737.0   
2608             42  ...                91112117.0   
2609              6  ...                56852018.0   

      UWHomes_TotalValue_AllTiers  UWHomes_MedianValue_AllTiers  \
2605                   82715412.0                      436490.5   
2606                   40365702.0                      131441.0   
2607                   50284001.0                      230495.0   
2608                  132505542.0                      385196.0   
2609                   59115506.0                      363355.0   

      AllHomes_Tier1_ShareUW  AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  \
2605                0.017241                0.061702                0.273834   
2606                0.213262                0.182137                0.095318   
2607                0.249315                0.210106                0.092910   
2608                0.069044                0.031627                0.097778   
2609                0.001339                0.001322                0.038366   

      AllHomes_AllTiers_ShareUW  UWHomes_ShareInTier1  UWHomes_ShareInTier2  \
2605                   0.120533              0.046512              0.168605   
2606                   0.162131              0.425000              0.371429   
2607                   0.180870              0.437500              0.379808   
2608                   0.066316              0.343396              0.158491   
2609                   0.010041              0.057971              0.043478   

      UWHomes_ShareInTier3  
2605              0.784884  
2606              0.203571  
2607              0.182692  
2608              0.498113  
2609              0.898551  

[5 rows x 24 columns]

Looking at this data, I think it really has four levels that describe the numbers. The columns `RegionType`, `RegionName`, `StateName`, and `MSAName` define unique rows of data. Again, note that these values can be missing. For example, row 2606 is Zip, 98595, Washington, but no `MSAName`. Why? MSA stands for Metropolitan Statistical Area and cover multiple zip codes.

Let’s think about the other variable definitions a bit. Zillow places homes into three tiers based on the home’s estimated market value, with Tier3 being the highest value. AllHomes refers to all homes in that tier or across all tiers. UWHomes refers to homes that are at risk for flooding. Note that there are some variables that are counts, some that are dollar values, and others that are percentages.



## Filtering our data

Let's try **filtering** our data and keeping just the MSA observations. The text calls this **complex selection**. See pg. 132.

Here's some [help on filtering](https://datagy.io/filter-pandas/) data with `pandas`.

In [5]:
uw[uw['RegionType'] == 'MSA'].head(15)

RegionType               RegionName      StateName  \
251        MSA             Aberdeen, WA     Washington   
252        MSA              Astoria, OR         Oregon   
253        MSA  Egg Harbor Township, NJ     New Jersey   
254        MSA            Baltimore, MD       Maryland   
255        MSA        East Falmouth, MA  Massachusetts   
256        MSA          Baton Rouge, LA      Louisiana   
257        MSA             Bay City, TX          Texas   
258        MSA             Beaumont, TX          Texas   
259        MSA           Bellingham, WA     Washington   
260        MSA               Boston, MA  Massachusetts   
261        MSA            Bremerton, WA     Washington   
262        MSA           Bridgeport, CT    Connecticut   
263        MSA          Brownsville, TX          Texas   
264        MSA            Brunswick, GA        Georgia   
265        MSA            Cambridge, MD       Maryland   

                     MSAName  AllHomes_Tier1  AllHomes_Tier2  AllHomes_Tier3  \
251             Aberdeen, WA         11606.0         11751.0         11749.0   
252              Astoria, OR          4745.0          5228.0          5568.0   
253  Egg Harbor Township, NJ         34185.0         34117.0         34088.0   
254            Baltimore, MD        298318.0        312603.0        316502.0   
255        East Falmouth, MA         45345.0         45358.0         45258.0   
256          Baton Rouge, LA         78815.0         79581.0         78731.0   
257             Bay City, TX          4161.0          4252.0          4030.0   
258             Beaumont, TX         49842.0         47829.0         44616.0   
259           Bellingham, WA         29239.0         30413.0         30956.0   
260               Boston, MA        409592.0        409966.0        410140.0   
261            Bremerton, WA         27865.0         27856.0         28119.0   
262           Bridgeport, CT         84535.0         84952.0         85028.0   
263          Brownsville, TX         47084.0         37642.0         35503.0   
264            Brunswick, GA         14827.0         14885.0         15532.0   
265            Cambridge, MD          6193.0          6251.0          6118.0   

     AllHomes_AllTiers  UWHomes_Tier1  UWHomes_Tier2  ...  \
251            35106.0         4119.0           2450  ...   
252            15541.0          811.0            376  ...   
253           102390.0         9116.0           9575  ...   
254           927423.0         2293.0           2910  ...   
255           135961.0         1651.0           1395  ...   
256           237127.0         2214.0           1904  ...   
257            12443.0          195.0            351  ...   
258           142287.0         1182.0           1043  ...   
259            90608.0         2033.0            812  ...   
260          1229698.0        15948.0          15210  ...   
261            83840.0          206.0            245  ...   
262           254515.0         2326.0           1895  ...   
263           120229.0         1621.0           1430  ...   
264            45244.0         2532.0           2924  ...   
265            18562.0          757.0           1119  ...   

     UWHomes_TotalValue_Tier3  UWHomes_TotalValue_AllTiers  \
251              1.734113e+08                 7.791588e+08   
252              9.138231e+07                 3.620604e+08   
253              5.576646e+09                 1.030631e+10   
254              3.698818e+09                 5.153664e+09   
255              4.205784e+09                 5.349099e+09   
256              5.392566e+08                 1.219648e+09   
257              1.175738e+08                 1.964820e+08   
258              1.466850e+08                 3.249066e+08   
259              7.000724e+08                 1.184436e+09   
260              2.851453e+10                 4.266696e+10   
261              2.076948e+09                 2.275970e+09   
262              6.041231e+09                 7.812512e+09   
263        

How about just North Carolina MSAs? I'll use the bigger DataFrame and just show the observations, rather than create a new DataFrame.

In [7]:
uw[(uw['RegionType'] == 'MSA') & (uw['StateName'] == 'North Carolina')].head(15)

RegionType            RegionName       StateName               MSAName  \
274        MSA    Elizabeth City, NC  North Carolina    Elizabeth City, NC   
277        MSA        Greenville, NC  North Carolina        Greenville, NC   
288        MSA      Jacksonville, NC  North Carolina      Jacksonville, NC   
292        MSA  Kill Devil Hills, NC  North Carolina  Kill Devil Hills, NC   
297        MSA     Morehead City, NC  North Carolina     Morehead City, NC   
302        MSA          New Bern, NC  North Carolina          New Bern, NC   
334        MSA        Washington, NC  North Carolina        Washington, NC   
336        MSA        Wilmington, NC  North Carolina        Wilmington, NC   

     AllHomes_Tier1  AllHomes_Tier2  AllHomes_Tier3  AllHomes_AllTiers  \
274          5887.0          5882.0          5868.0            17637.0   
277         22042.0         19844.0         18836.0            60722.0   
288         17623.0         19413.0         17989.0            55025.0   
292          3194.0          3213.0          3486.0             9893.0   
297         13555.0         13843.0         12927.0            40325.0   
302         13637.0         13409.0         12440.0            39486.0   
334          1289.0          1069.0          1043.0             3401.0   
336         37100.0         40034.0         37475.0           114609.0   

     UWHomes_Tier1  UWHomes_Tier2  ...  UWHomes_TotalValue_Tier3  \
274         2198.0           1364  ...              4.784827e+08   
277           71.0             34  ...              6.786239e+06   
288          792.0           1123  ...              6.370269e+08   
292         1461.0           1584  ...              9.128899e+08   
297         3195.0           3308  ...              2.316937e+09   
302         1364.0           1138  ...              5.442292e+08   
334          255.0            201  ...              5.513269e+07   
336         3367.0           3339  ...              4.265842e+09   

     UWHomes_TotalValue_AllTiers  UWHomes_MedianValue_AllTiers  \
274                 8.833370e+08                      138246.0   
277                 1.776378e+07                       99179.5   
288                 1.029168e+09                      264070.5   
292                 1.650730e+09                      280409.0   
297                 3.449861e+09                      244691.0   
302                 8.553945e+08                      163643.0   
334                 9.990351e+07                      116902.5   
336                 6.761845e+09                      383067.0   

     AllHomes_Tier1_ShareUW  AllHomes_Tier2_ShareUW  AllHomes_Tier3_ShareUW  \
274                0.373365                0.231894                0.283401   
277                0.003221                0.001713                0.001221   
288                0.044941                0.057848                0.083329   
292                0.457420                0.492997                0.542456   
297                0.235706                0.238966                0.359712   
302                0.100022                0.084868                0.132235   
334                0.197828                0.188026                0.182167   
336                0.090755                0.083404                0.127578   

     AllHomes_AllTiers_ShareUW  UWHomes_ShareInTier1  UWHomes_ShareInTier2  \
274                   0.296252              0.420670              0.261053   
277                   0.002108              0.554688              0.265625   
288                   0.062045              0.231986              0.328940   
292                   0.498939              0.295989              0.320908   
297                   0.276578              0.286470              0.296602   
302                   0.105025              0.328912              0.274415   
334                   0.189944              0.394737              0.311146   
336                   0.100228              0.293114              0.290676   

     UWHome

Notice how I needed to put `()` around each condition. Python is testing to see if each of these are `True` or `False` and then filtering our data accordingly. Pgs. 132 - 135 of our text show you more about this type of **boolean** filtering.